validator/sawtooth_validator/execution/context_manager.py

```diff
@@ -210,112 +272,120 @@ def commit_context(self, context_id_list, virtual):
                 "MerkleRoots not all equal, yet asking to merge")
 
         merkle_root = self._contexts[first_id].merkle_root
-        tree = MerkleDatabase(self._database, merkle_root)
-
-        merged_updates = {}
+        tree = MerkleDatabase(self._database, merkle_root=merkle_root)
+        updates = dict()
         for c_id in context_id_list:
-            with self._shared_lock:
-                context = self._contexts[c_id]
-                del self._contexts[c_id]
-            for k in context.get_writable_address_value_dict().keys():
-                if k in merged_updates:
+            context = self._contexts[c_id]
+            for add in context.get_state().keys():
+                if add in updates:
                     raise CommitException(
-                        "Duplicate address {} in context {}".format(k, c_id))
-            merged_updates.update(context.get_writable_address_value_dict())
+                        "Duplicate address {} in context {}".format(
+                            add, c_id))
 
-        new_root = merkle_root
+            effective_updates = {}
+            for k, val_fut in context.get_state().items():
+                value = val_fut.result()
+                if value is not None:
+                    effective_updates[k] = value
 
-        add_value_dict = {}
-        for k, val_fut in merged_updates.items():
-            value = val_fut.result()
-            if value is not None:
-                add_value_dict[k] = value
+            updates.update(effective_updates)
 
-        new_root = tree.update(set_items=add_value_dict, virtual=virtual)
+        state_hash = tree.update(updates, virtual=False)
+        # clean up all contexts that are involved in being squashed.
+        base_c_ids = []
+        for i in context_id_list:
```

Comment:
Good call. I will change it to something short but more descriptive. id is a python builtin (https://docs.python.org/3/library/functions.html#id) and so variable names shouldn't shadow builtins.